### ISSUE TITLE
HDR Tone Mapping Is Still Causing Clipping in 2021

### ISSUE THREAD

**Issue by sovon** [mpv-player/mpv#8849]
**MacOS Big Sur 11.3.1 (20E241) - Apple Silicon Mac Mini M1 - LG 4K HDR 27UK850**
```
mpv 0.33.1 Copyright © 2000-2020 mpv/MPlayer/mplayer2 projects
 built on Mon Apr  5 10:16:25 PDT 2021
FFmpeg library versions:
   libavutil       56.51.100 (runtime 56.70.100)
   libavcodec      58.91.100 (runtime 58.134.100)
   libavformat     58.45.100 (runtime 58.76.100)
   libswscale      5.7.100 (runtime 5.9.100)
   libavfilter     7.85.100 (runtime 7.110.100)
   libswresample   3.7.100 (runtime 3.9.100)
FFmpeg version: 4.4
```

I've talked about this issue in the past. Not going to waste everyone's time by doing this same thing again. Just wanted to let the community know about a recent discovery of mine. I purchased a license of Infuse today after I saw what that app can do with HDR content. I want to support this mpv open source community and so I'm posting these discoveries of mine after hours of testing in case they can be implemented in MPV as well.

It's a well known fact that displaying HDR content on a SDR monitor is never a straightforward conversion. I have played around A LOT with the 6 algorithms - hable, clim, gamma, linear, mobius and reinhard. And for the past 5 years my mpv config has consistently looked something like this:

```
#----------------------------------------------------------- VIDEO
icc-profile-auto
hwdec=videotoolbox
tone-mapping=reinhard
tone-mapping-param=0.65

#----------------------------------------------------------- PLAYER
no-osd-bar
volume=100
volume-max=150
window-scale=0.5
script-opts=osc-scalefullscreen=0.7

#----------------------------------------------------------- SUBTITLE
sub-scale=0.5
sub-color="#dadada"


#----------------------------------------------------------- EXTERNAL
--cache=yes
--demuxer-max-bytes=50M


#----------------------------------------------------------- SCREENSHOT
screenshot-directory=/Users/perfectidiot/Desktop
```

_These settings have changed here and there over time but never have I ever seen ANY player that handled this scene from Avengers:IW this good like Infuse does here. That out of the way, here are the comparisons between Infuse and MPV. Take a look:_

**MPV:** Obvious clipping aside you'll notice that the overall image is too orange-ish/yellow-ish
![MPVV](https://user-images.githubusercontent.com/18725800/119231658-a7f33600-bb3f-11eb-8c50-1da88df1878c.JPG)

**Infuse:** No clipping anywhere that I could find and more impressively the image is very natural looking. I've matched it with OTT stream of disneyplus and the stream looked almost identical to this, so I think whatever Infuse is doing here, whatever magic algorithm/tweak they have concocted, it's doing the job right. Please compare all these images side by side to see the orangish hue.
![infusee](https://user-images.githubusercontent.com/18725800/119231661-a9bcf980-bb3f-11eb-80fc-cbd3dc548239.JPG)

**MPV**
![MPV (4k-hdr)](https://user-images.githubusercontent.com/18725800/119231663-ac1f5380-bb3f-11eb-833b-d39e9eda8152.JPG)

**Infuse**
![Infuse (4k-hdr)](https://user-images.githubusercontent.com/18725800/119231667-ad508080-bb3f-11eb-92bf-00898f8b6f65.JPG)

**MPV**
![MPV](https://user-images.githubusercontent.com/18725800/119232309-60ba7480-bb42-11eb-8318-0fdc92c025bc.JPG)

**Infuse** 
![Infuse](https://user-images.githubusercontent.com/18725800/119232310-61530b00-bb42-11eb-9506-bac035684c6f.JPG)

**MPV:** On dark scenes, you'll notice that the image from mpv looks a little less sharp, bit washed out and less contrasty compared to Infuse.
![mpvv](https://user-images.githubusercontent.com/18725800/119233029-bee85700-bb44-11eb-9687-760b401cdd34.JPG)

**Infuse** 
![infusee](https://user-images.githubusercontent.com/18725800/119233033-c0b21a80-bb44-11eb-90e5-d3459e0b86ae.JPG)

Instead of Infuse, I'll continue to use MPV as my primary player cause it's lightweight, way faster and customisable. If anyone can help me achieve the same level of HDR results as my Infuse app, I'd be very grateful.

**Comment by qyot27** [comment]
https://github.com/mpv-player/mpv/commit/c9f6c458ea54f595127684a3bcd94f578ec42b1f?

**Comment by haasn** [comment]
You might be interested in https://code.videolan.org/videolan/libplacebo/-/merge_requests/155

**Comment by sovon** [comment]
> [c9f6c45](https://github.com/mpv-player/mpv/commit/c9f6c458ea54f595127684a3bcd94f578ec42b1f)?

Apologies but I'm a code illiterate. I'm a designer and figuring out how to write the config files and how to change between various options in MPV were about the most difficult thing with respect to coding(?) I had to do in years. I wouldn't know how to read what you've linked, let alone what to do with it.

**Comment by sovon** [comment]
> You might be interested in https://code.videolan.org/videolan/libplacebo/-/merge_requests/155
> 
> Or rather, if you can help test it, it'd be a great way to fast-track those changes into libplacebo (and then mpv, via #8799)

I can put the effort and spend time testing whatever is needed but I don't think I understand the technicality all that much. Is there something easier that I can follow?

**Comment by Argon-** [comment]
> > [c9f6c45](https://github.com/mpv-player/mpv/commit/c9f6c458ea54f595127684a3bcd94f578ec42b1f)?
> 
> Apologies but I'm a code illiterate. I'm a designer and figuring out how to write the config files and how to change between various options in MPV were about the most difficult thing I had to do in years. I wouldn't know how to read what you've linked, let alone figuring out what to do with it.

He's asking whether your version of mpv contains this piece of code. The precise version of your mpv binary would allow to deduce this.

**Comment by sovon** [comment]
> > > [c9f6c45](https://github.com/mpv-player/mpv/commit/c9f6c458ea54f595127684a3bcd94f578ec42b1f)?
> > 
> > 
> > Apologies but I'm a code illiterate. I'm a designer and figuring out how to write the config files and how to change between various options in MPV were about the most difficult thing I had to do in years. I wouldn't know how to read what you've linked, let alone figuring out what to do with it.
> 
> He's (presumably) asking whether your version of mpv contains this piece of code. The precise version of your mpv binary would allow to deduce this.

Oh. I just added my MPV build information on the top of the opening post.

**Comment by Argon-** [comment]
According to this info your mpv contains the mentioned commit and you could try with `tone-mapping=bt.2390` if you didn't already.

**Comment by sovon** [comment]
I didn't before but I just did now and the results are dim, literally.

**The Dimmest**
```
icc-profile-auto
hwdec=videotoolbox
tone-mapping=bt.2390
tone-mapping-param=0.65
```
![Screenshot 2021-05-22 at 11 05 10 PM](https://user-images.githubusercontent.com/18725800/119235960-8fd8e200-bb52-11eb-9feb-3a1f2acf19ef.JPG)

**Medium Dim**
```
hwdec=videotoolbox
tone-mapping=bt.2390
tone-mapping-param=0.65
```
![Screenshot 2021-05-22 at 11 06 58 PM](https://user-images.githubusercontent.com/18725800/119235967-98c9b380-bb52-11eb-8dbe-c53b29a6e87b.JPG)

**Infuse**
![infff](https://user-images.githubusercontent.com/18725800/119236013-c282da80-bb52-11eb-907f-fedf0052bf83.JPG)

**Comment by NSQY** [comment]
https://mpv.io/manual/master/#options-gamut-clipping might be the problem in this case, I have found that along with bt.2390/hdr-compute peak chroma can get verrry clipped.

**Comment by haasn** [comment]
@NSQY oh yes indeed, I just discovered that same bug in libplacebo but haven't fixed it for vo_gpu yet, 1 sec

**Comment by haasn** [comment]
Fixed the issue I mentioned but on second though I'm not convinced it's actually relevant for you, since it only triggered with HDR outputs (not SDR outputs).

**Comment by Robot-DaneelOlivaw** [comment]
FWIW, [`tone-mapping-desaturate`](https://mpv.io/manual/master/#options-tone-mapping-desaturate) and [`tone-mapping-desaturate-exponent`](https://mpv.io/manual/master/#options-tone-mapping-desaturate-exponent) also controls desaturation for overly bright colors.

**Comment by sovon** [comment]
> FWIW, [`tone-mapping-desaturate`](https://mpv.io/manual/master/#options-tone-mapping-desaturate) and [`tone-mapping-desaturate-exponent`](https://mpv.io/manual/master/#options-tone-mapping-desaturate-exponent) also control desaturation for overly bright colors.

This is from two years ago but may still be relevant. May be.
Please see this: https://github.com/mpv-player/mpv/issues/6405#issuecomment-491594394

**Comment by NSQY** [comment]
> Fixed the issue I mentioned but on second though I'm not convinced it's actually relevant for you, since it only triggered with HDR outputs (not SDR outputs).

I may have been conflating problems.

With sources that have a strong chrominance component, gamut clipping can get a little bizarre https://slow.pics/c/8Mvmb77e

**Comment by Doofussy2** [comment]
Try disabling compute peak. `hdr-compute-peak=no`

**Comment by Doofussy2** [comment]
I wonder if lowering the `--target-peak` might help, here? The default is 203, and the display may not be that high?

Try `--target-peak=190`

**Comment by sovon** [comment]
> I wonder if lowering the `--target-peak` might help, here? The default is 203, and the display may not be that high?
> 
> Try `--target-peak=190`
The higher the number, the dimmer the image.
```
hwdec=videotoolbox
tone-mapping=reinhard
tone-mapping-param=0.65
target-peak=300
```
![Screenshot 2021-05-23 at 2 23 41 AM](https://user-images.githubusercontent.com/18725800/119240569-1222cf80-bb6e-11eb-9488-37a3620e16c0.JPG)

```
target-peak=190
```
![Screenshot 2021-05-23 at 2 21 13 AM](https://user-images.githubusercontent.com/18725800/119240579-22d34580-bb6e-11eb-92b2-2a4221c42f30.JPG)

```
target-peak=130
```
![Screenshot 2021-05-23 at 2 22 19 AM](https://user-images.githubusercontent.com/18725800/119240584-2c5cad80-bb6e-11eb-8438-cb34e89eb707.JPG)

**Comment by Doofussy2** [comment]
> The higher the number, the dimmer the image.

Yes, that's why I suggested lowering it. I should have been clear in intending you to try the default and not reinhard. bt.2390 is the default and the default peak is 203. So try bt.2390 with a lower target peak. And remember that bt.2390 isn't tunable so it doesn't have a mapping parameter.

So all you need to do to test, is use one line: `target-peak=190` or whatever number you wish to test. This is just to see if you attain the brightness that you are looking for. You could also test it with `gamut-clipping=no`.

**Comment by sovon** [comment]
```
hwdec=videotoolbox
tone-mapping=bt.2390
target-peak=150
gamut-clipping=no
```

In my opening post I mentioned 2 issues - one was clipping and the other was orange-ish hue. WIth the above setting, the orange-ish hue is almost gone while target-peak=150 is set to achieve approximate brightness. The clipping however is still present.

**MPV - image from the opening post** w/ orange-ish hue (reinhard + tone-mapping param 0.65)
![119231663-ac1f5380-bb3f-11eb-833b-d39e9eda8152](https://user-images.githubusercontent.com/18725800/119241229-79db1980-bb72-11eb-9719-06793c2a30b9.JPG)

**MPV -bt.2390** - orange-ish hue gone but clipping is there
![Screenshot 2021-05-23 at 2 50 40 AM](https://user-images.githubusercontent.com/18725800/119241238-94ad8e00-bb72-11eb-9a1b-936633a6d83d.JPG)

**Infuse**
![119231667-ad508080-bb3f-11eb-92bf-00898f8b6f65](https://user-images.githubusercontent.com/18725800/119241249-a2631380-bb72-11eb-8045-26e3dbe1bb30.JPG)

**Comment by Doofussy2** [comment]
Now you'll have to test haasn's new commit addressing the gamut clipping.

**Comment by haasn** [comment]
@sovon What you might actually want is to add this

```
tone-mapping-desaturate=1.0
tone-mapping-desaturate-exponent=0.0
```

**Comment by sovon** [comment]
Like https://github.com/mpv-player/mpv/issues/6405#issuecomment-491594394 tone-mapping-desaturate is blowing the highlights.

```
hwdec=videotoolbox
tone-mapping=bt.2390
target-peak=150
gamut-clipping=no
tone-mapping-desaturate=1.0
tone-mapping-desaturate-exponent=0.0
```
**MPV:** blue-purple becomes white-ish purple and highlight data is lost.
![MPV](https://user-images.githubusercontent.com/18725800/119252374-7cb52900-bbc9-11eb-9d38-2edfa186e2af.JPG)

```
hwdec=videotoolbox
tone-mapping=bt.2390
target-peak=150
gamut-clipping=no
###tone-mapping-desaturate=1.0
tone-mapping-desaturate-exponent=0.0
```
**MPV:** Only with tone-mapping-desaturate-exponent=0.0 the results look just a bit better, less highlight data is lost but the  shift in the purple isn't as drastic as the one above.
![tone-mapping-desaturate-exponent=0 0](https://user-images.githubusercontent.com/18725800/119252441-e8979180-bbc9-11eb-9a89-16aac3f00c80.JPG)

**Infuse:** As a reference.
![Infuse](https://user-images.githubusercontent.com/18725800/119252444-edf4dc00-bbc9-11eb-8bea-79f78b0e35cc.JPG)

These are all of course done in my existing mpv version that comes from `brew upgrade` and not that commit @Doofussy2 had mentioned. I am not sure how to do that.

**Comment by Doofussy2** [comment]
So here's something out of left field. With:

```
hdr-compute-peak=no
brightness=1
tone-mapping-desaturate-exponent=0.7
gamma=2
```

I get this:

![mpv-shot0013](https://user-images.githubusercontent.com/15950919/119273598-7ec6ba00-bbc0-11eb-818f-b8d35d709f68.jpg)



That's pretty damn close to the Infuse shot, except for the increased flaring in the center.

I have noticed there is darkening in all things HDR. Whether I tone-map or passthrough, the image is always slightly darker than it should be. For all content, I have a general correction of `brightness=1` and `gamma=2`. I think there must be a gamma variation. 

@haasn it looks to me like that what you're working on with the blackpoint and not setting a gamma curve, would correct this?

**Comment by sovon** [comment]
I may be wrong but in my personal experience I have NEVER been able to get a satisfied result by tweaking brightness or contrast. It may work on a movie I'm watching at that moment but it will definitely ruin other films for me later on. With brightness increased by even just 1 unit, I can clearly tell  where the letterbox starts, and I dark scenes never come out as good. Same thing happens for contrast as well, +/- whatever that is.

**Comment by Doofussy2** [comment]
Don't adjust the contrast.

**Comment by sovon** [comment]
I don't do either. I leave brightness, contrast, gamma & saturation untouched and advise others to do the same. To me these settings may help in unique circumstances such as where the display panel itself isn't as good or where the ICC color profile isn't great or has a limited LAB plot (enhancing the look and feel by bumping up the saturation a bit etc. is only temporary) But that is VERY rare and I never meddle with superficial tweaks to keep it as close as to the filmmaker's vision as possible.

In order to get a reference to the "filmmaker's vision", I take screenshots from bluray.com and compare my image to those for that exact frame. Sadly these OTT platforms like netflix or apple TV doesn't allow me to take screenshots so I have to compare windows but I've found that the effort is worth it.

For brightness and contrast, I found out the issue the hard way. I was watching Mandalorian; when Mando was in that spider cave, MPV looked so dim compared to the safari version of disneyplus even though reinhard, param-0.65 was there. I modified the brightness contrast and gamma of MPV to match the safari window and it was great for the rest of the season but a week later I was watching Riddick and omg I've never seen a washed out pitch black scene like that before. Because I have seen the film in theatre and remember it was actually pitch black, at first I thought my monitor broke but then I saw the mpv config and figured out why it was like that.

**Comment by Doofussy2** [comment]
Yes, that's where blackpoint/gamma/trc comes in. Which is the PR that haasn linked to and why I made reference to it. 

https://code.videolan.org/videolan/libplacebo/-/merge_requests/155

Unfortunately, my build environment doesn't allow me to apply patches, or I would be testing the work with libplacebo.  Hopefully haasn will complete it in the not too distant future. I'm very much looking forward to putting it through it's paces.

**Comment by Robot-DaneelOlivaw** [comment]
I think it's hard to judge whether a tone-mapped screenshot is dim or not without knowing each others' screen brightness. If you care about brightness precision, you might want to measure peak brightness of your screen and set `target-peak` accordingly.

**Comment by Doofussy2** [comment]
> I think it's hard to judge whether a tone-mapped screenshot is dim or not without knowing each others' screen brightness.

All you have to do is put both images next to each other on the same display. Which ostensibly, posting them on here, does. 

What I found was, when I juxtaposed the video using passthrough to my Samsung Q80R with how mpv tone-maps with the default settings, was that it was very similar (clipping not withstanding). The 'darkness' was comparative, which is what I would want. Why obtain HDR media if you just want to strip it and remake it the same as the SDR mastering? Just use the SDR version, and you'll have exactly what you want. I *want* the deep color and the expanded light range. If you don't want HDR, don't use HDR. So once the blackpoint is correctly applied with haasn's PR, I think we'll be golden.

**Comment by sovon** [comment]
> I think it's hard to judge whether a tone-mapped screenshot is dim or not without knowing each others' screen brightness.

You are right. Independently across different devices that comparison would definitely be weird. But if you open two images of the exact same frame and toggle between them with your keyboard maybe, the differences would be drastically pronounced. From there it's only a matter of choice - which one you _prefer_ to the other. As a reference we can compare frames from established forums or from sources that we know are sure to produce accurate results (such as bluray.com screenshot/video quality review section or maybe from a web OTT service such as Apple TV or DIsneyPlus or HBO Max etc.). So you can compare your rendered image with theirs and easily find out what is different in your media consumption device.

> Why obtain HDR media if you just want to strip it and remake it the same as the SDR mastering? Just use the SDR version, and you'll have exactly what you want.

Well, I for one would gladly rent an SDR master for sure because I don't have a HDR TV. I, like many others, do all my media consumption on my desktop computer which although has a 4K _HDR_ 4K monitor, that hdr functionality remains mostly unused. Implementation of HDR in macOS or Windows is abysmal to say the least and no one can work on hdr mode because the gamut gets destructively compressed (at least for designers who export sRGB for web). That would be madness. So many people would just use the SDR mode of their monitor.

Also they don't make 4K UHD blurays SDR mastered. Every single UHD bluray I ever saw is HDR mastered. And I personally am lazy to rip it myself and store the 90GB data on my storage, so from time to time, I might obtain a small 4K encode from a popular release group in the internet, which 99% of the time is an HDR HEVC encode.

But I think at this point I need to make the point again that I didn't raise this issue here in github for my personal use. Infuse is doing wonderfully although it costs a little too much but I am happy with what I have. My intention was to show it to the community, so that the same quality / features can be more publicly and openly available. An ideal media player should be able to handle the weirdness of the media file, irrespective of source. An end user who might not be aware of all these nitty gritties, would still be able to consume a piece of content in its full glory (or as close to the filmmaker's vision as possible).

**Comment by Doofussy2** [comment]
>An ideal media player should be able to handle the weirdness of the media file, irrespective of source. An end user who might not be aware of all these nitty gritties, should ideally still be able to consume a piece of content in its full glory (or as close to the filmmaker's vision as possible).

Aside from the clipping, mpv does tone-map the way I believe it should. It should closely mimic HDR, as best it can, and not try to reproduce SDR. That was my point. 

And I have a lot of 4K SDR media. It's very common.
![Bright](https://user-images.githubusercontent.com/15950919/119308399-d9dcc900-bc21-11eb-9708-511ec6345aef.jpg)

**Comment by amariami** [comment]
> **MPV:** On dark scenes, you'll notice that the image from mpv looks a little less sharp, bit washed out and less contrasty compared to Infuse.

Adjusting gamma?
 [`--target-trc=gamma2.4`](https://mpv.io/manual/master/#options-target-trc)
 [`--vf=format=gamma=gamma2.4`](https://mpv.io/manual/master/#video-filters-%3Cgamma%3E)

Image attachment just for reference
![SPOILER_Screenshot 2021-05-25 034532](https://user-images.githubusercontent.com/79769464/119485358-d0c82680-bd0b-11eb-8d80-5b6e63b9e2a3.png)

**Comment by hooke007** [comment]
> vf=format=gamma=gamma2.4

Do not do this if you watch HDR videos and use icc profile at the same time. Just to ruin the color.

> target-trc=gamma2.4

This option will be overrided by icc.

**Comment by amariami** [comment]
> 
> This option will be overridden by ICC.

Thanks, I gotcha

I read a reference from here 
[`BT.1886`](https://nick-shaw.github.io/cinematiccolor/common-rgb-color-spaces.html#x34-1350004.1.2)
and
[`Compact ICC Profiles from github saucecontrol`](https://github.com/saucecontrol/Compact-ICC-Profiles)

**Comment by amariami** [comment]
Set the 3D-LUT size `--icc-3dlut-size=<r>x<g>x<b>` could solve the problem?
`--icc-3dlut-size=256x256x256   #2 to 512`
`--scaler-lut-size=10   #4 to 10`

I see reference in here;
--audio-resample-filter-size=32
--audio-resample-cutoff=0.0
#--audio-resample-phase-shift=30 #polyphase entries. (..., 10->1024, 11->2048, 12->4096, ...) (default: 10->1024)
--audio-format=float
--audio-samplerate=48000
#--audio-swresample-o-add=resampler=soxr 
#--audio-swresample-o-add=precision=28
#--audio-swresample-o-add=matrix_encoding=dplii
#--audio-swresample-o-add=matrix_encoding=dolby
#--audio-swresample-o-add=matrix_encoding=none
#--audio-swresample-o-add=rematrix_maxval=1.0
###see https://ffmpeg.org/ffmpeg-resampler.html#Resampler-Options
--af-add=lavfi=[aresample=resampler=soxr:precision=28:cutoff=0]
--af-add=lavfi=[stereotools=phasel=true:phaser=true:phase=360:bmode_out=power] #set phase degree 90/180/270/360
--af-add=lavfi=[asoftclip=type=quintic:threshold=1:output=1:oversample=16] ##type set to "hard" "tanh" "atan" "cubic" "exp" "alg" "quintic" "sin" "erf"

**Comment by sovon** [comment]
Is there anything else I can test out? (any new development to this perhaps)

**Comment by haasn** [comment]
@sovon You could help test out [libplacebo](https://code.videolan.org/videolan/libplacebo) master, which got a complete overhaul of the HDR code very recently.

Those changes will make their way into mpv itself with #8799.

**Comment by sovon** [comment]
> @sovon You could help test out [libplacebo](https://code.videolan.org/videolan/libplacebo) master, which got a complete overhaul of the HDR code very recently.

How do I test it with mpv?

**Comment by Doofussy2** [comment]
We need a kind person to make a test build for us 😁 Sadly, until it's merged, I can't build with it.

**Comment by sovon** [comment]
With my previous configuration, every time I played a SDR film in mpv, I used to get a warning, something like - "icc profile... too high... sanity.. etc". With the current `tone-mapping=bt.2390` I don't see that anymore.
Anyways, just chiming in. Infuse is high maintenance af, not suitable at all for click and play. For personal use too, I hope mpv gets this new update soon you guys have been teasing about.
